Feature/data platform infra doc update (#358)
amido/amido.github.io: docs/workloads/azure/data/getting_started/core_data_platform_deployment_azure.md

description: Infrastructure deployment
keywords:
  - stacks cli
  - data
  - infrastructure
  - azure
  - template
---

This section provides an overview of generating a new Data Platform project and deploying core infrastructure for Stacks Data Platform.
This aligns to the workflow shown in the [deployment architecture](../architecture/architecture_data_azure.md#data-engineering-workloads) section.

> It assumes all prerequisites are in place, including:

Reviewer: ```suggestion
It assumes all [prerequisites](../requirements_data_azure.md#azure) are in place, including:
```

> * [Azure](../requirements_data_azure.md#azure)
> * [Azure Pipelines variable groups](../requirements_data_azure.md#azure-pipelines-variable-groups)

Reviewer: ```suggestion
* Azure subscription and service principal
* Azure DevOps project with [Pipelines variable groups](../requirements_data_azure.md#azure-pipelines-variable-groups)
* A remote git repository for hosting the generated project (this guide assumes `main` is the primary branch in this repo)
```

Download and install the `stacks-cli` using [Stacks CLI](../../../../stackscli/about) page. Please refer to the **Stacks.CLI.Manual** in the latest `stacks-cli` release for detailed instruction.

To construct a Data Platform project, two primary cli commands are required: `stacks-cli interactive` and `stacks-cli scaffold`.

The interactive command is designed to ask questions on the command line about the configuration
required for setting up Ensono Digital Stacks. It will then save this configuration out to a file that can be
read in using the scaffold command.

```cmd
stacks-cli interactive
```

The majority of the questions are self-explanatory; please refer to the **Stacks.CLI.Manual** for further detail, however the following two will define the type of the target project. 

| Question                                      | Required value for data project |
|-----------------------------------------------|---------------------------------|
| What framework should be used for the project?| infra                           |
| Which type of infrastructure is required?     | data                            |

> The resulting configuration file named stacks.yml contains all of the configuration that was used to generate the projects,

Reviewer: ```suggestion
The resulting configuration file named `stacks.yml` contains all of the configuration that was used to generate the project,
```

which means it can be used to produce the same project stack again.

The CLI can be used with a configuration file to generate the Ensono Digital Stacks based projects using `stacks-cli scaffold`.

```cmd
stacks-cli scaffold -c ./stacks.yml
```

> Open the project locally and push the generated project to target repository in `main/master` branch.

Reviewer: Updated based on adding an assumption above that `main` is used in the new repo.
```suggestion
Open the project locally and push the generated project to the target remote repository's `main` branch.
```

> ## Step 2: Add Infrastructure pipeline in Azure Devops

Reviewer: ```suggestion
## Step 2: Add Infrastructure pipeline in Azure DevOps
```

Open the project locally and create a new feature branch e.g.:

```bash
git checkout -b feat/infra-pipeline
```

YAML file containing a template Azure DevOps CI/CD pipeline for building and deploying the core infrastructure, named `air-infrastructure-data.yml` is configured in the `build/azDevOps/azurefolder`

YAML file `air-infrastructure-data.yml` should be added as the definition for a new pipeline in Azure DevOps.

1. Sign-in to your Azure DevOps organization and go to your project
2. Go to Pipelines, and then select New pipeline
3. Name the new pipeline eg `amido.stacks-data-infrastructure`
4. For the pipeline definition, specify the YAML file in the project repository feature branch (e.g. `air-infrastructure-data.yml`) and save
5. The new pipeline will require access to any Azure DevOps pipeline variable groups specified in the pipeline YMAL. Under each variable group, go to 'Pipeline permissions' and add the pipeline.

> The `build/azDevOps/azure/network` folder has a YAML file entitled `air-infrastructure-data-network.yml` that contains a template Azure DevOps CI/CD pipeline for networking and self-hosted agent infrastructure. If the client/user does not have their own network and self-hosted agent, then we must establish the networking pipeline in Azure Devops following the procedures outlined above and it must before the core infrasturcture pipeline. The `build/azDevOps/azure/network` folder also contains a YAML file called `air-infrastructure-data-network-vars.yml` that contains the variables used in the networking pipeline. These variables must be updated as well as the project requirements.

Reviewer 2: ```suggestion
The `build/azDevOps/azure/network` folder contains a YAML file named `air-infrastructure-data-network.yml` that serves as a template for Azure DevOps CI/CD pipeline for networking and self-hosted agent infrastructure. If the client or user does not have their own network and self-hosted agent, they must establish the networking pipeline in Azure DevOps following the procedures outlined above. This setup should be completed before initiating the core infrastructure pipeline. The `build/azDevOps/azure/network` folder also includes another YAML file called `air-infrastructure-data-network-vars.yml` that contains the variables used in the networking pipeline. These variables must be updated as well as the project requirements.
```

> If the client/user have their own network and self hosted agent then networking pipeline wont be required.

Reviewer 2: ```suggestion
If the client/user has their own network and self-hosted agent, then the networking pipeline won't be required.
```

## Step 3: Deploy Infrastructure in non-production environment

> Run the  pipeline configured in Step 2

Reviewer: ```suggestion
Run the pipeline configured in Step 2 to commence the build and deployment process.
```

> Running this pipeline in Azure DevOps will deploy the artifacts into the non-production (nonprod) environment. If successful, the generated resources will now be available in the nonprod Stacks environment.

Reviewer: I would something here on:
 - Where you can monitor the progress and see if it's successful (i.e. within Pipelines)
 - Where you can view the resources that have been deployed (i.e. in the portal, what the resource group would be named)

Author: updated as per feedback

> 

Reviewer: We have some variable groups that are listed as being added after core infra deployment. If so, we need a step here to say they need adding.

Author: updated as per feedback

> 

Author: Yeah, I thinks its better to have this separate page.

## Step 4: Deploy Infrastructure in further environments 

> 

Reviewer: I would add something here along the lines of:

By default Stacks provides a framework for managing the platform across two environments - `nonprod` and `prod` - configured as below.

Author: updated as per feedback

```cmd
stacks-cli scaffold -c ./stacks.yml
```

Open the project locally and push the generated project to the target remote repository's `main` branch.

## Step 2: Add Infrastructure pipeline in Azure DevOps

Open the project locally and create a new feature branch e.g.:

```bash
git checkout -b feat/infra-pipeline
```

YAML file containing a template Azure DevOps CI/CD pipeline for building and deploying the core infrastructure, named `air-infrastructure-data.yml` is configured in the `build/azDevOps/azurefolder`

YAML file `air-infrastructure-data.yml` should be added as the definition for a new pipeline in Azure DevOps.

1. Sign-in to your Azure DevOps organization and go to your project
2. Go to Pipelines, and then select **New pipeline**
3. Name the new pipeline, e.g. `amido.stacks-data-infrastructure`
4. For the pipeline definition, specify the YAML file in the project repository feature branch (`air-infrastructure-data.yml`) and save
5. The new pipeline will require access to any Azure DevOps pipeline variable groups specified in the pipeline YMAL. Under each variable group, go to 'Pipeline permissions' and add the pipeline.

> The `build/azDevOps/azure/network` folder contains a YAML file named `air-infrastructure-data-network.yml` that serves as a template for Azure DevOps CI/CD pipeline for networking and self-hosted agent infrastructure. If the client or user does not have their own network and self-hosted agent, they must establish the networking pipeline in Azure DevOps following the procedures outlined above. This setup should be completed before initiating the core infrastructure pipeline. The `build/azDevOps/azure/network` folder also includes another YAML file called `air-infrastructure-data-network-vars.yml` that contains the variables used in the networking pipeline. These variables must be updated as well as the project requirements.

Reviewer: Could add a subheading before this part, e.g. 'Optional: Networking and self-hosted agent'

Author: updated

If the client/user has their own network and self-hosted agent, then the networking pipeline won't be required.

## Step 3: Deploy Infrastructure in non-production environment

Run the pipeline configured in Step 2 to commence the build and deployment process.

> Running this pipeline in Azure DevOps will initiate the deployment of artifacts into the non-production (nonprod) environment.It's important to monitor the progress of this deployment to ensure its success. You can track the progress and status of the deployment within the Pipelines section of Azure DevOps.

Reviewer: ```suggestion
Running this pipeline in Azure DevOps will initiate the deployment of artifacts into the non-production (nonprod) environment. It's important to monitor the progress of this deployment to ensure its success. You can track the progress and status of the deployment within the Pipelines section of Azure DevOps.
```

> If successful, the core infrasturcture resources will now be available in the nonprod Stacks environment. To view these deployed resources, navigate to the Azure portal and search for the resource group associated with the deployment. This resource group is named based values provided during step 1, such as company name, project name,region,component. Within the resource group, you'll find a list of the resources that were deployed.

Reviewer: ```suggestion
If successful, the core infrasturcture resources will now be available in the nonprod Stacks environment. To view these deployed resources, navigate to the Azure portal (https://portal.azure.com/) and search for the resource group associated with the deployment. This resource group is named based upon values provided during step 1 in the pattern
`companyname-projectname-stage-region-component` (for example: `amido-stacks-dev-euw-de`). Within the resource group, you'll find a list of the resources that were deployed.
```

> Name Format for Resource Groups  - companyname-projectname-stage-region-component eg amido-stacks-dev-euw-de

Reviewer: Can remove this if using the suggestion above.

Author: updated

> Once core infrasturcture resources are deployed in nonprod environment,please add the required variable values in `amido-stacks-de-pipeline-nonprod` variable group. For additional information, see [Pipelines variable groups](../requirements_data_azure.md#azure-pipelines-variable-groups).

Reviewer: You could also add a subheader here, e.g. 'Update variable group values'

```suggestion
Once core infrasturcture resources are deployed in nonprod environment, values will need adding into the nonprod variable group to reflect the deployed resources (e.g. `amido-stacks-de-pipeline-nonprod`). For additional information, see [Pipelines variable groups](../requirements_data_azure.md#azure-pipelines-variable-groups).
```

## Step 4: Deploy Infrastructure in further environments

* Deployment to the non-production (nonprod) environment is triggered on a feature branch when a pull request is open
* Deployment to the production (prod) environment is triggered on merging to the `main` branch, followed by manual approval of the release step.

> Once core infrasturcture resources are deployed in nonprod environment,please add the required variable values in `amido-stacks-de-pipeline-prod` variable group.For additional information, see [Pipelines variable groups](../requirements_data_azure.md#azure-pipelines-variable-groups).

Reviewer: You could match the suggested wording and layout for nonprod here.

By default Stacks provides a framework for managing the platform across two environments - nonprod and prod.

> The template CI/CD pipelines provided are based upon these two platform environments (nonprod and prod) - but these may be amended depending upon the specific requirements of your project and organisation.

Reviewer: I'd move this part to the start of step 4

Author: updated
